Client/Frontend/Browser/PasswordHelper.swift

```diff
@@ -171,4 +172,95 @@ class PasswordHelper: BrowserHelper {
+                        completion(password: Password(credential: credential!, protectionSpace: challenge.protectionSpace))
+                        return
+                    }
+                } else {
+                    credential = nil
+                }
+            } else {
+                credential = nil
+            }
+        }
+
+        // If we have some credentials, we'll show a prompt with them.
+        if let credential = credential {
+            println("Use guess")
+            let password = Password(credential: credential, protectionSpace: challenge.protectionSpace)
+            promptForUsernamePassword(viewController, password: password, completion: completion)
+        } else {
+            // Otherwise, try to look some up
+            let options = QueryOptions(filter: challenge.protectionSpace.host, filterType: .None, sort: .None)
+            profile.passwords.get(options, complete: { (cursor) -> Void in
+                var password = cursor[0] as? Password
+                println("Found \(password?.username)")
+                if password == nil {
+                    password = Password(credential: nil, protectionSpace: challenge.protectionSpace)
+                }
+                self.promptForUsernamePassword(viewController, password: password!, completion: completion)
+            })
+        }
+    }
+
+    private func promptForUsernamePassword(viewController: UIViewController, password: Password, completion: (password: Password?) -> Void) {
+
+        // Show an alert controller (prompt) to get the username and password.
+        let alert = UIAlertController(title: NSLocalizedString("Authentication required", comment: "Authentication prompt title"),
+            message: NSLocalizedString(NSString(format: "A username and password are being requested by %@. The site says: %@",
+                password.hostname ?? "(unknown)",
+                password.httpRealm ?? "(unknown)") as String,
+                comment: "Authentication prompt message"),
+            preferredStyle: UIAlertControllerStyle.Alert)
+
+        // Add a login button.
+        let action = UIAlertAction(title: NSLocalizedString("Login", comment: "Authentication prompt login button"),
+            style: UIAlertActionStyle.Default) { (action) -> Void in
+                let user = (alert.textFields?[0] as! UITextField).text
+                let pass = (alert.textFields?[1] as! UITextField).text
+
+                let credential = NSURLCredential(user: user, password: pass, persistence: .ForSession)
+                completion(password: Password(credential: credential, protectionSpace: password.protectionSpace))
+                self.setPassword(password)
+        }
+        alert.addAction(action)
+
+        // Add a cancel button.
+        let cancel = UIAlertAction(title: NSLocalizedString("Cancel", comment: "Authentication prompt cancel button"), style: UIAlertActionStyle.Cancel) { (action) -> Void in
```

Comment:
Nit: IMO, it's generally more readable to put these `NSLocalizedString`s in a variable on their own line so these don't expand so much.